### PR TITLE
ibmcloud: e2e-test cases for ibmcloud to check peer-pod Logs User from container

### PR DESCRIPTION
**Pull request by sudharshanibm3**
Added e2e-test cases for ibmcloud to check peer-pod Logs User from container

Fixes: #1001 

Note: Current PR is dependent on #942

### CODE REVIEW

**Review by stevenhorsman — COMMENTED** [review]
Hey Sudharshan, can you remove the jobs test commit from this PR please as it isn't related to this issue? Also I think you need a commit description and DCO sign off in the workflow commit. Thanks!

**Review by stevenhorsman — COMMENTED** [review]
On first glance it looks okay, but I think there is a lot of overlap in code with #942, so I think we need to get that merged first and then rebase on it otherwise we'll likely have a lot of conflicts

**Comment by sudharshanibm3** [comment]
Hi @stevenhorsman ,
Rebased the code with latest staging branch

File: .github/workflows/test-images.yaml [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+jobs:
+  list-dockerfiles:
+    runs-on: ubuntu-latest
+    if: ${{ github.event_name == 'workflow_dispatch' }}
+    outputs:
+      matrix: ${{ steps.set-matrix.outputs.matrix }}
+    steps:
+      - uses: actions/checkout@v2
+      - id: set-matrix
+        run: echo "::set-output name=matrix::$(ls test/e2e/fixtures/Dockerfile.* | jq -R -s -c 'split("\n")[:-1]')"
```
Comment (stevenhorsman) [inline .github/workflows/test-images.yaml:18]: Just to check - is `jq` installed on `ubuntu-latest`?

Comment (stevenhorsman) [reply .github/workflows/test-images.yaml:18]: Any chance you can kick this off manually in your fork to check this? (though I appreciate that you might need to update the registry for that.

Comment (stevenhorsman) [reply .github/workflows/test-images.yaml:18]: Also - it looks like `set-output` is deprecated now: https://github.blog/changelog/2022-10-11-github-actions-deprecating-save-state-and-set-output-commands/

**Comment by sudharshanibm3** [comment]
Hi @stevenhorsman ,
Here I ran the [workflow](https://github.com/sudharshanibm3/cloud-api-adaptor/actions/runs/5179007906) in my forked repo with my [registry](https://quay.io/repository/sudharshanibm3/test-images?tab=tags)
Screenshot 2023-06-05 at 9 02 37 PM

File: .github/workflows/test-images.yaml (hunk 1 above)
Comment (stevenhorsman) [reply .github/workflows/test-images.yaml]: Thanks for the fixes!

File: .github/workflows/test-images.yaml [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+  build:
+    name: Create Test Images
+    needs: list-dockerfiles
+    runs-on: ubuntu-latest
+    strategy:
+      fail-fast: false
+      matrix:
+        targets: ${{ fromJson(needs.list-dockerfiles.outputs.matrix) }}
```
[the remark this answers is not in the document]
Comment (sudharshanibm3) [reply .github/workflows/test-images.yaml:25]: I think it's due to the inability in unmarshalling property of the matrix
here I found something related to that, [solution](https://github.com/nektos/act/issues/927#issuecomment-1035013048)

File: .github/workflows/test-images.yaml [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+        tags: |
+          quay.io/${{ secrets.QUAY_USERNAME }}/test-images:${{env.DOCKER_TAG}}
+        push: true
+        context: .
+        platforms: linux/s390x,linux/amd64
+        file: ${{matrix.targets}}
+        build-args: |
+          "ARCH=${{ matrix.arch }}"
+          "UBUNTU_IMAGE_URL="
+          "UBUNTU_IMAGE_CHECKSUM="
```
Comment (stevenhorsman) [inline .github/workflows/test-images.yaml:70]: Why are these arguments needed?

Comment (sudharshanibm3) [reply .github/workflows/test-images.yaml]: Sorry, I think these arguments are not required. Removed the arguments in updated code

File: test/e2e/common_suite_test.go [hunk 4]
```diff
@@ -236,6 +254,39 @@ func newTestCase(t *testing.T, testName string, assert CloudAssert, assessMessag
 
 	return testCase
 }
+func ComparePodLogString(ctx context.Context, t *testing.T, client klient.Client, Pod v1.Pod, ExpectedPodlogString string) (bool, string, error) {
```
Comment (stevenhorsman) [inline test/e2e/common_suite_test.go:257]: A new line between functions makes it easier to read IMHO

File: test/e2e/common_suite_test.go [hunk 5]
```diff
@@ -236,6 +254,39 @@ func newTestCase(t *testing.T, testName string, assert CloudAssert, assessMessag
 
 	return testCase
 }
```
Comment (stevenhorsman) [inline test/e2e/common_suite_test.go:256]: A new line between functions makes it easier to read IMHO

File: test/e2e/common_suite_test.go (hunk 4 above)
Comment (stevenhorsman) [inline test/e2e/common_suite_test.go:257]: Convention is for go variables to start with lower case letters
```suggestion
func ComparePodLogString(ctx context.Context, t *testing.T, client klient.Client, pod v1.Pod, expectedPodlogString string) (bool, string, error) {
```

Comment (stevenhorsman) [inline test/e2e/common_suite_test.go:257]: Is this function intended to be exported? otherwise I think it should start with a lower case too?

File: test/e2e/common_suite_test.go [hunk 6]
```diff
@@ -114,15 +112,58 @@ func (tc *testCase) run() {
...
 					t.Fatal(err)
 				}
-				if err = wait.For(conditions.New(client.Resources()).PodRunning(tc.pod), wait.WithTimeout(WAIT_POD_RUNNING_TIMEOUT)); err != nil {
-					t.Fatal(err)
+				if tc.expectedPodLogString != "" {
+					if err = wait.For(conditions.New(client.Resources()).PodPhaseMatch(tc.pod, v1.PodSucceeded), wait.WithTimeout(WAIT_POD_RUNNING_TIMEOUT)); err != nil {
+						t.Fatal(err)
+					}
+				} else {
+					if err = wait.For(conditions.New(client.Resources()).PodRunning(tc.pod), wait.WithTimeout(WAIT_POD_RUNNING_TIMEOUT)); err != nil {
```
Comment (stevenhorsman) [inline test/e2e/common_suite_test.go:120]: Quick question - do the custom test images go into/end on a different state to the others e.g. nginx and that's why we need to do this branch?

File: test/e2e/common_suite_test.go [hunk 7]
```diff
@@ -114,15 +112,58 @@ func (tc *testCase) run() {
...
+					}
 				}
+
 			}
 			return ctx
 		}).
 		Assess(tc.assessMessage, func(ctx context.Context, t *testing.T, cfg *envconf.Config) context.Context {
 			client := cfg.Client()
 			var podlist v1.PodList
+			if tc.pod != nil {
```
Comment (stevenhorsman) [inline test/e2e/common_suite_test.go:131]: Any particularly reason that you moved the `if tc.pod != nil` logic up above the `if tc.job != nil` branch in this PR?

Comment (sudharshanibm3) [reply test/e2e/common_suite_test.go:131]: No It wont affect the logic of the code.  Placement can be changed if required

File: test/e2e/common_suite_test.go (hunk 6 above)
Comment (sudharshanibm3) [reply test/e2e/common_suite_test.go:120]: Yeah 
nginx image will be in `Running` state
our custom image will go to `Completed` state 
so made a separate branch for tracking pods with `Completed` state

**Comment by sudharshanibm3** [comment]
Hi @stevenhorsman ,
Updated the requested changes

File: test/e2e/common_suite_test.go (hunk 6 above)
Comment (stevenhorsman) [reply test/e2e/common_suite_test.go:122]: I guess my concern here is that this branch isn't really based on whether we have expected logging, but rather what the image is, so we could get into a situation where we want to check the logs of an image that stays in running. I wonder if instead of branching on this if we could update the wait to look for either running or completed and then it should be valid for all cases?

File: test/e2e/common_suite_test.go (hunk 7 above)
Comment (stevenhorsman) [reply test/e2e/common_suite_test.go:133]: Yes - I'd rather keep the original placement as it makes the diff in the PR smaller, thanks

**Comment by stevenhorsman** [comment]
> Hi @stevenhorsman , Updated the requested changes

Thanks, I've added comments on a couple of them.

BTW @mattarnoatibm - if you have any cycles to review this before your vacation then I'd definitely value your opinion!

File: test/e2e/common_suite_test.go (hunk 6 above)
Comment (sudharshanibm3) [reply test/e2e/common_suite_test.go:128]: Let me know if this solution works

- Decalred a `podState` in testCase struct
- If any pod requires different state to assess rather than `Running`, they can declare via `withCustomPodState`
- So users can declare different podState like `Succeeded`, `Pending`,`Unknown`,`Failed` and podState will dynamically add those states while setup part

Implemented above approach in updated code, please let me know your views on this approach

File: test/e2e/common_suite_test.go [hunk 8]
```diff
@@ -50,6 +50,7 @@ type testCase struct {
 	job                  *batchv1.Job
 	testCommands         []testCommand
 	expectedPodLogString string
+	podState             string
```
Comment (mattarnoatibm) [inline test/e2e/common_suite_test.go:53]: Can we make the type of `podState` a `v1.podPhase`?

File: test/e2e/common_suite_test.go [hunk 9]
```diff
@@ -114,15 +120,58 @@ func (tc *testCase) run() {
 				if err = client.Resources().Create(ctx, tc.pod); err != nil {
 					t.Fatal(err)
 				}
-				if err = wait.For(conditions.New(client.Resources()).PodRunning(tc.pod), wait.WithTimeout(WAIT_POD_RUNNING_TIMEOUT)); err != nil {
-					t.Fatal(err)
+				if tc.podState != "" {
```
Comment (mattarnoatibm) [inline test/e2e/common_suite_test.go:123]: `wait.For(conditions.New(client.Resources()).PodRunning(tc.pod), wait.WithTimeout(WAIT_POD_RUNNING_TIMEOUT))` should be equivalent to `wait.For(conditions.New(client.Resources()).PodPhaseMatch(tc.pod, v1.PodRunning)`, so I think we can remove the need to have this `if` `else` branch by

- Setting a default `podState` of `v1.PodRunning` in `newTestCase`
- Change the `wait.For` code to `wait.For(conditions.New(client.Resources()).PodPhaseMatch(tc.pod, tc.podState), wait.WithTimeout(WAIT_POD_RUNNING_TIMEOUT))`, which covers test cases where the test expects a pod to reach the running phase and where the test expects a pod to reach the Succeeded phase.

File: test/e2e/common_suite_test.go [hunk 10]
```diff
@@ -114,15 +120,58 @@ func (tc *testCase) run() {
...
+					isLogStringValid, LogString, err := comparePodLogString(ctx, t, client, *tc.pod, tc.expectedPodLogString)
+					if err != nil {
+						t.Fatal(err)
+					}
+					if isLogStringValid {
+						t.Logf("Log output of peer pod:%s", LogString)
+					} else {
+						t.Errorf("Invalid Log output from peer pod:%s", LogString)
+					}
+				} else if tc.testCommands != nil {
```
Comment (mattarnoatibm) [inline test/e2e/common_suite_test.go:153]: I don't think having checking logs and running a command inside a pod container in exclusive branches is going to be valid for all test cases we could write, because you could write a test case where we want to check the logs in a pod that reaches the Running phase, and then run a command inside a container in that running pod.

I suggest checking if `tc.podState` is `v1.PodRunning` and running the test commands if that condition is true, and making that conditional separate from the `tc.expectedPodLogString != ""` check.

File: test/e2e/common_suite_test.go [hunk 11]
```diff
@@ -237,6 +263,40 @@ func newTestCase(t *testing.T, testName string, assert CloudAssert, assessMessag
 	return testCase
 }
 
+func comparePodLogString(ctx context.Context, t *testing.T, client klient.Client, customPod v1.Pod, expectedPodlogString string) (bool, string, error) {
```
Comment (mattarnoatibm) [inline test/e2e/common_suite_test.go:266]: I don't see `t *testing.T` used anywhere in this function.

File: test/e2e/common_suite_test.go [hunk 12]
```diff
@@ -237,6 +263,40 @@ func newTestCase(t *testing.T, testName string, assert CloudAssert, assessMessag
...
+				_, err = io.Copy(buf, podLogs)
+				if err != nil {
+					return
+				}
+				podLogString = strings.TrimSpace(buf.String())
+			}()
+		}
+	}
+	if podLogString != expectedPodlogString {
+		return false, podLogString, nil
```
Comment (mattarnoatibm) [inline test/e2e/common_suite_test.go:295]: Instead of returning `false` and a `nil` you can return a  new error (e.g., from `errors.New()`) here.

This also lets you simplify the types returned by  `comparePodLogString` to just `(string, error)`.

File: test/e2e/fixtures/Dockerfile.testuser [hunk 13]
```diff
@@ -0,0 +1,4 @@
+FROM --platform="${TARGETPLATFORM}" alpine:latest
+RUN addgroup -S othergroup && adduser -S otheruser -G othergroup
+USER otheruser
+ENTRYPOINT [ "/bin/sh", "-c", "whoami" ]
```
Comment (mattarnoatibm) [inline test/e2e/fixtures/Dockerfile.testuser:4]: NP: Missing final newline.

File: test/e2e/ibmcloud_test.go [hunk 14]
```diff
@@ -92,6 +92,13 @@ func TestCreatePeerPodWithJob(t *testing.T) {
 	doTestCreatePeerPodWithJob(t, assert)
 }
 
+func TestCreatePeerPodWithUserImage(t *testing.T) {
```
Comment (mattarnoatibm) [inline test/e2e/ibmcloud_test.go:95]: I think the name of this test case and the name of the `doTestCreatePeerPodWithUserImage` function should be different. I think the purpose of this test case it to test fetching logs from a pod container. The fact the pod container image is a user-built (or test case built) image is, I believe, circumstantial. So I think a name like `TestCreatePeerPodAndCheckLogs`.

**Review by mattarnoatibm — CHANGES_REQUESTED** [review]
Thanks for contributing and refining these tests.

I've added some comments and suggestions.

File: test/e2e/ibmcloud_test.go (hunk 14 above)
Comment (sudharshanibm3) [reply test/e2e/ibmcloud_test.go:95]: Hi @mattarnoatibm ,
There are test case which  need to be merged after this PR, which checks the logs of workdir, envvariable, etc
So can we keep this name as `TestCreatePeerPodAndCheckUserLogs`, `TestCreatePeerPodAndCheckWorkDirectoryLogs`...  ?

Comment (mattarnoatibm) [reply test/e2e/ibmcloud_test.go:95]: Suggestion - since this test case checks the output of a container image entrypoint command, rename the test case to `TestCreatePeerPodAndCheckOutputOfEntrypointCmd`?

**Comment by sudharshanibm3** [comment]
Hi @mattarnoatibm ,
Updated code with requested changes

File: test/e2e/common_suite_test.go (hunk 7 above)
Comment (stevenhorsman) [reply test/e2e/common_suite_test.go:135]: I think we are still waiting for this fix?

File: test/e2e/ibmcloud_test.go (hunk 14 above)
Comment (stevenhorsman) [reply test/e2e/ibmcloud_test.go]: So I'm not sure that this works fully like this Sudharshan? We have a number of tests that do this behaviour, but they will use different images and expect different logged output, so I think whilst we could have a basic test called `TestCreatePeerPodAndCheckOutputOfEntrypointCmd`, we would need to pass in the pod name, image, expected log, testcase name, description etc to have it unique for the 3/4 different test cases and at that point I'm not sure if it's messier to have a shared function with ~7/8 parameters rather than individual ones?

File: test/e2e/common_suite_test.go (hunk 7 above)
Comment (sudharshanibm3) [reply test/e2e/common_suite_test.go:159]: Exchanged `job` and `pod` statements in updated code

File: test/e2e/ibmcloud_test.go (hunk 14 above)
Comment (stevenhorsman) [reply test/e2e/ibmcloud_test.go]: I'm not a huge fan of using the map to get around there being a lot of parameters. I guess maybe the best thing to do it go back to the `TestCreatePeerPodAndCheckUserLogs` name and then look at refactoring it once we add more tests with common code, rather than attempting to prematurely refactor for the future?

Comment (sudharshanibm3) [reply test/e2e/ibmcloud_test.go]: Hi @stevenhorsman ,
Reverted the test name to `TestCreatePeerPodAndCheckUserLogs`

Comment (stevenhorsman) [reply test/e2e/ibmcloud_test.go]: It still looks to be named `doTestCreatePeerPodAndCheckOutputOfEntrypointCmd`? Anyway, the main thing is to remove the map/extra parameters and review that once we add the env/workdir test

**Review by stevenhorsman — APPROVED** [review]
This looks good to me - as mentioned we can work on further refactoring once the subsequent related changes are ready

File: test/e2e/common_suite_test.go [hunk 15]
```diff
@@ -237,6 +255,40 @@ func newTestCase(t *testing.T, testName string, assert CloudAssert, assessMessag
...
+				_, err = io.Copy(buf, podLogs)
+				if err != nil {
+					return
+				}
+				podLogString = strings.TrimSpace(buf.String())
+			}()
+		}
+	}
+	if podLogString != expectedPodlogString {
+		return podLogString, errors.New(err.Error())
```
Comment (mattarnoatibm) [inline test/e2e/common_suite_test.go:287]: This will throw a nil pointer dereference if nothing in the inner function returns an error but the pod container logs from the pod are not equal to the pod container logs the test case expected.

I'd suggest creating a new error object with a new custom error message string.

I think you need a separate check before this check to return `podLogString, err` if `err` is not nil.

File: test/e2e/common_suite_test.go [hunk 16]
```diff
@@ -292,7 +344,7 @@ func getSuccessfulAndErroredPods(ctx context.Context, t *testing.T, client klien
 func doTestCreateSimplePod(t *testing.T, assert CloudAssert) {
 	namespace := envconf.RandomName("default", 7)
 	pod := newNginxPod(namespace)
-	newTestCase(t, "SimplePeerPod", assert, "PodVM is created").withPod(pod).run()
+	newTestCase(t, "SimplePeerPod", assert, "PodVM is created").withPod(pod).withCustomPodState(v1.PodRunning).run()
```
Comment (mattarnoatibm) [inline test/e2e/common_suite_test.go:347]: I'd make the `newTestCase` function add a default pod state `v1.PodRunning` to the test case so you don't have to use `withCustomPodState(v1.PodRunning)` when creating every test case.

File: test/e2e/common_suite_test.go [hunk 17]
```diff
@@ -237,6 +255,40 @@ func newTestCase(t *testing.T, testName string, assert CloudAssert, assessMessag
 	return testCase
 }
 
+func comparePodLogString(ctx context.Context, client klient.Client, customPod v1.Pod, expectedPodlogString string) (string, error) {
+	podLogString := ""
+	var podlist v1.PodList
+	clientset, err := kubernetes.NewForConfig(client.RESTConfig())
+	if err != nil {
+		return podLogString, errors.New(err.Error())
```
Comment (mattarnoatibm) [inline test/e2e/common_suite_test.go:263]:
```suggestion
		return podLogString, err
```

File: test/e2e/common_suite_test.go [hunk 18]
```diff
@@ -237,6 +255,40 @@ func newTestCase(t *testing.T, testName string, assert CloudAssert, assessMessag
...
+	}
+	if err := client.Resources(customPod.Namespace).List(ctx, &podlist); err != nil {
+		return podLogString, errors.New(err.Error())
```
Comment (mattarnoatibm) [inline test/e2e/common_suite_test.go:266]:
```suggestion
		return podLogString, err

**Review by mattarnoatibm — CHANGES_REQUESTED** [review]
Thanks for the changes.

I've got a few more changes to request that fall outside of general refactoring.

**Review by mattarnoatibm — CHANGES_REQUESTED** [review]

File: test/e2e/common_suite_test.go [hunk 19]
```diff
@@ -232,11 +250,48 @@ func newTestCase(t *testing.T, testName string, assert CloudAssert, assessMessag
...
+				_, err = io.Copy(buf, podLogs)
+				if err != nil {
+					return
+				}
+				podLogString = strings.TrimSpace(buf.String())
+			}()
+		}
+	}
+	if podLogString != expectedPodlogString {
+		if err != nil {
```
[the remark this answers is not in the document]
Comment (sudharshanibm3) [reply test/e2e/common_suite_test.go:288]: Hi @mattarnoatibm ,
I think updated code resolves this error

[the remark this answers is not in the document]
Comment (mattarnoatibm) [reply test/e2e/common_suite_test.go:288]: I think there's a corner case where `expectedPodlogString` is `""` and `err` is not nil where the `comparePodLogString` function will return `podLogString, nil` when it should return `podLogString, err`. So you could catch that by taking the `if err != nil {` case outside of the `if podLogString != expectedPodlogString {` case.

[the remark this answers is not in the document]
Comment (sudharshanibm3) [reply test/e2e/common_suite_test.go:288]: ok, then added an extra condition in updated code. Could you please check if that's ok?

**Review by mattarnoatibm — APPROVED** [review]
LGTM. Thanks for the contribution!
